cli/cmd/component-delete.go

```diff
@@ -0,0 +1,170 @@
+}
+
+func runDelete(cmd *cobra.Command, args []string) {
+	contextLogger := log.WithFields(log.Fields{
+		"command": "lokoctl component delete",
+		"args":    args,
+	})
+
+	lokoCfg, diags := getLokoConfig()
+	if len(diags) > 0 {
+		contextLogger.Fatal(diags)
+	}
+
+	componentsToDelete := make([]string, len(args))
+	copy(componentsToDelete, args)
+
+	if len(args) == 0 {
+		componentsToDelete = make([]string, len(lokoCfg.RootConfig.Components))
+
+		for i, component := range lokoCfg.RootConfig.Components {
+			componentsToDelete[i] = component.Name
+		}
+	}
+
+	// Verify if all the components user has specified exist
+	for _, componentName := range componentsToDelete {
+		_, err := components.Get(componentName)
+		if err != nil {
+			contextLogger.Fatal(err)
+		}
+	}
+
+	if !askForConfirmation(
+		fmt.Sprintf(
+			"The following components will be deleted:\n\t%s\n\nAre you sure you want to proceed?",
+			strings.Join(componentsToDelete, "\n\t"),
+		),
+	) {
+		contextLogger.Info("Components deletion cancelled.")
+		return
+	}
+
+	kubeconfig, err := getKubeconfig()
+	if err != nil {
+		contextLogger.Fatalf("Error in finding kubeconfig file: %s", err)
+	}
+
+	if err := deleteComponents(lokoCfg, kubeconfig, componentsToDelete...); err != nil {
+		contextLogger.Fatal(err)
+	}
+}
+
+func deleteComponents(lokoCfg *config.Config, kubeconfig string, componentNames ...string) error {
+	for _, componentName := range componentNames {
+		fmt.Printf("Deleting component '%s'...\n", componentName)
+
+		// This is already verified, so we can ignore the error
+		component, _ := components.Get(componentName)
+
+		componentConfigBody := lokoCfg.LoadComponentConfigBody(componentName)
+
+		if diags := component.LoadConfig(componentConfigBody, lokoCfg.EvalContext); len(diags) > 0 {
+			return diags
+		}
+
+		if err := deleteComponent(componentName, component, kubeconfig, deleteNamespace); err != nil {
+			return err
+		}
+
+		fmt.Printf("Successfully deleted component %q!\n", componentName)
+	}
+
+	// Add a line to distinguish between info logs and errors, if any.
+	fmt.Println()
+
+	return nil
+}
+
+// deleteComponent deletes a component.
+func deleteComponent(name string, c components.Component, kubeconfig string, deleteNSBool bool) error {
+	ns := c.Metadata().Namespace
+	if ns == "" {
+		// This should never fail in real user usage, if this does that means the component was not
+		// created with all the needed information
+		panic(fmt.Errorf("component %s namespace is empty", name))
+	}
+
+	cfg, err := util.HelmActionConfig(ns, kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed preparing helm client: %w", err)
+	}
+
+	uninstall := action.NewUninstall(cfg)
+
+	resp, err := uninstall.Run(name)
+	// ignore the err when we have deleted the release already or it does not exist for some reason
+	if err != nil && !strings.Contains(err.Error(), "uninstall: Release not loaded") {
```

Comment:
Maybe we could check if the release exists before trying to uninstall it to avoid doing string comparison on errors. This seems rather fragile.

Reply:
done, but even when we check if the installation/release exists. It all boils down on what assumptions we are willing to trust more?

Reply:
IMO it's better to check if the release exists, as this is the same what `helm` is doing and it's a GET operation, which is OK if it fails, where DELETE is something which should not fail normally. Also with GET there is specific error type as far as I remember.